### PR TITLE
update to 2.9.2

### LINKED ISSUE #38: Windows builds link wrong openssl
This has broken sometime between #35 & #36 and is currently plaguing #37. The pytest runner shows that the wrong version is being used (instead of 1.1.1g in the environment)

```
============================= test session starts =============================
platform win32 -- Python 3.8.2, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
OpenSSL: OpenSSL 1.1.1  11 Sep 2018
rootdir: %SRC_DIR%
plugins: hypothesis-5.10.3
collected 107077 items / 11 skipped / 107066 selected

tests\test_cryptography_utils.py ....                                    [  0%]
[...]
```

This was only caught by accident because the cryptography testsuite contains a test that [fails](https://github.com/conda-forge/cryptography-feedstock/pull/37#issuecomment-612525838) on openssl < 1.1.1b. 
```
=========================== short test summary info ===========================
FAILED tests/hazmat/primitives/test_ed448.py::TestEd448Signing::test_malleability
== 1 failed, 103852 passed, 3235 skipped, 626 warnings in 347.38s (0:05:47) ===
```

### PR DESCRIPTION
**Pull request by h-vetinari**
Checklist
* [x] Used a [fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

Somehow, this [release](https://github.com/pyca/cryptography/releases) (more than a week ago) didn't get picked up by the autotick-bot. Also, the windows builds should be fixed after having been merged in a broken state in #36

Edit: Assuming that something similar to the presently implemented `run_test.py` gets merged, this will fix #38

### CODE REVIEW

**Comment by h-vetinari** [comment]
@conda-forge-admin, please rerender

**Comment by h-vetinari** [comment]
One failure on windows:

```
=========================== short test summary info ===========================
FAILED tests/hazmat/primitives/test_ed448.py::TestEd448Signing::test_malleability
== 1 failed, 103852 passed, 3235 skipped, 626 warnings in 347.38s (0:05:47) ===
```

<details>

```
================================== FAILURES ===================================
_____________________ TestEd448Signing.test_malleability ______________________

self = <tests.hazmat.primitives.test_ed448.TestEd448Signing object at 0x0000026A2CD12F98>
backend = <cryptography.hazmat.backends.openssl.backend.Backend object at 0x0000026A709F9B38>

    def test_malleability(self, backend):
        # This is a signature where r > the group order. It should be
        # rejected to prevent signature malleability issues. This test can
        # be removed when wycheproof grows ed448 vectors
        public_bytes = binascii.unhexlify(
            "fedb02a658d74990244d9d10cf338e977565cbbda6b24c716829ed6ee1e4f28cf"
            "2620c052db8d878f6243bffc22242816c1aaa67d2f3603600"
        )
        key = Ed448PublicKey.from_public_bytes(public_bytes)
        with pytest.raises(InvalidSignature):
>           key.verify(signature, b"8")
E           Failed: DID NOT RAISE <class 'cryptography.exceptions.InvalidSignature'>
```

</details>

@conda-forge/cryptography
Could this be related to using different compilers? At least the upstream tests [did not fail](https://github.com/pyca/cryptography/runs/556114929).

@alex @reaperhulk
Is this something that you've seen before? Should I open an issue upstream?

**Comment by alex** [comment]
This looks like what you'd get if you patched cryptography to try to enable ed448 on an earlier version of openssl than what we support.

**Comment by reaperhulk** [comment]
Yeah OpenSSL 1.1.1b fixed a malleability bug in ed448 and ed25519 and this test verifies against that. Should skip on OpenSSL < 1.1.1b, but there are a few edges where it's possible to have it think it's supported (e.g. compiled on newer, run on older in dynamic linking scenarios).

**Comment by h-vetinari** [comment]
@alex @reaperhulk
Thanks a lot for the responses! Actually, the build is using `1.1.1f` (or at least trying to), but there may be some linking problem:
```
WARNING (cryptography): run-exports library package conda-forge::openssl-1.1.1f-hfa6e2cd_0 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
```

**Comment by h-vetinari** [comment]
From looking at [`_get_openssl_libraries`](https://github.com/pyca/cryptography/blob/master/src/_cffi_src/build_openssl.py#L18) (and how it's passed through [`_cffi_src/utils.py`](https://github.com/pyca/cryptography/blob/master/src/_cffi_src/utils.py)), I'm guessing the `cryptography` build is somehow picking up an outdated system-library, instead of the one from conda.

`libssl` and `libcrypto` are built by the openssl-feedstock, but somehow their path seems to have lower priority? While trying to figure this out on my local windows machine, I ran into https://github.com/conda-forge/openssl-feedstock/issues/54, which *might* be related.

In any case: CC @conda-forge/openssl

**Comment by h-vetinari** [comment]
Ping @conda-forge/core. Since everyone needs `cryptography`, unblocking it is quite important IMO.

**Comment by h-vetinari** [comment]
@beckermr 
Thanks for taking a look. Which status page? The one from [conda-forge](https://conda-forge.org/status/)? Or somewhere within the worker [bot](https://circleci.com/gh/regro/circle_worker)? I've never had to look this up before.

That solves one problem (why the upgrade bot didn't trigger a PR), but it doesn't fix that the wrong openssl is linked in...

**Comment by h-vetinari** [comment]
@conda-forge-admin, please rerender

**Comment by beckermr** [comment]
Yea the conda forge one. There is a version update status section. It is brand new which is why you have not seen it. And yeah it doesn’t solve the openssl issue.

**Comment by h-vetinari** [comment]
The new status page (esp. with the interactive migration dashboards) is very cool!

**Comment by beckermr** [comment]
@conda-forge/core Can someone who knows the openssl stack look at this? Maybe @jjhelmus?

**Comment by jjhelmus** [comment]
The [cryptography docs](https://cryptography.io/en/latest/installation/#building-cryptography-on-windows) mention setting the `LIB` and `INCLUDE` variables to point at the location where OpenSSL is installed.  I believe `%LIBRARY_BIN%` and `%LIBRARY_INC%` would be the correct locations for the OpenSSL install provided by the conda package.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -10,7 +10,10 @@ source:
 
 build:
   number: 0
-  script: "{{ PYTHON }} -m pip install . -vv"
+  script:
+    - set LIB=%LIBRARY_BIN%;%LIB%   # [win]
+    - set INCLUDE=%LIBRARY_INC%;%INCLUDE%   # [win]
```
Comment (isuruf) [inline recipe/meta.yaml:15]: This is done automatically by the compiler activation script

**Comment by h-vetinari** [comment]
@jjhelmus 
Thanks for looking into this. The build still fails on windows however (travis is unrelated: full disk on worker).

**Comment by jjhelmus** [comment]
This is the link command split by whitespace
```
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\link.exe 
/nologo 
/INCREMENTAL:NO 
/LTCG 
/DLL 
/MANIFEST:EMBED,ID=2 
/MANIFESTUAC:NO 
/LIBPATH:D:\bld\cryptography_1587056801933\_h_env\libs 
/LIBPATH:D:\bld\cryptography_1587056801933\_h_env\PCbuild\amd64 
/LIBPATH:D:\bld\cryptography_1587056801933\_h_env\Library\bin "
/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\ATLMFC\lib\x64" 
"/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\lib\x64" 
"/LIBPATH:C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\lib\um\x64" 
"/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.17763.0\ucrt\x64" 
"/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.17763.0\um\x64" 
/LIBPATH:D:\bld\cryptography_1587056801933\_h_env\Library\lib 
/LIBPATH:D:\bld\cryptography_1587056801933\_h_env\Library\lib 
libssl.lib 
libcrypto.lib 
advapi32.lib 
crypt32.lib 
gdi32.lib 
user32.lib 
ws2_32.lib 
/EXPORT:PyInit__openssl build\temp.win-amd64-3.8\Release\build\temp.win-amd64-3.8\Release\_openssl.obj 
/OUT:build\lib.win-amd64-3.8\cryptography\hazmat\bindings\_openssl.cp38-win_amd64.pyd 
/IMPLIB:build\temp.win-amd64-3.8\Release\build\temp.win-amd64-3.8\Release\_openssl.cp38-win_amd64.lib 
/NXCOMPAT 
/DYNAMICBASE
```

The static openssl libraries are used (libssl.lib and libcrypto.lib) which should be comning from the last two LIBPATH entries.  It would be better to dynamically link these if upstream supports this so that updates to OpenSSL do not requirement a rebuild of cryptography.

It would be helpful to see what version of OpenSSL is linked.  The output of:
```
python -c "from cryptography.hazmat.backends.openssl import backend; print(backend.openssl_version_text())"
```
Should provide this.

**Comment by jjhelmus** [comment]
`OpenSSL: OpenSSL 1.1.1  11 Sep 2018`.  That is not the version from the conda package.

**Comment by h-vetinari** [comment]
> @jjhelmus It would be better to dynamically link these if upstream supports this so that updates to OpenSSL do not requirement a rebuild of cryptography.

@alex @reaperhulk 
Is dynamic linking on windows supported upstream? The [installation docs](https://cryptography.io/en/latest/installation/#building-cryptography-on-windows) don't mention dynamic linking except in the context of macOs.

> @jjhelmus: `OpenSSL: OpenSSL 1.1.1 11 Sep 2018`. That is not the version from the conda package.

Yeah, we came to the [same conclusion](https://github.com/conda-forge/cryptography-feedstock/pull/37#issuecomment-612576023) based on the fact that the failing test could only fail with an OpenSSL < 1.1.1b

**Comment by reaperhulk** [comment]
Dynamic linking will work fine. We statically link in our infrastructure since we ship wheels to users who have no consistent availability of libraries, but conda has a different set of constraints there. All of our linux distro packagers use dynamic linking as well.

**Comment by isuruf** [comment]
Let's ask @mingwandroid. Azure installs openssl dlls in `C:\Windows\System32`. Is there a way to make sure our binaries look in `%LIBRARY_BIN%` before that directory?

**Comment by mingwandroid** [comment]
Yes. Set CONDA_DLL_SEARCH_MODIFICATION_ENABLE..

**Comment by h-vetinari** [comment]
> @mingwandroid: Yes. Set CONDA_DLL_SEARCH_MODIFICATION_ENABLE.

Thanks, I set it to `1` as mentioned in [these docs](https://docs.conda.io/projects/conda/en/latest/user-guide/troubleshooting.html) I found!

I just added it to the build script for now, but I'm wondering: will this percolate to a user install, or would those still falsely link to a system-dll (if present)? In that case, we'd be passing the tests here, but still creating vulnerable environments.

I'm guessing this has to go into the environment activation scripts?

**Comment by h-vetinari** [comment]
@mingwandroid 
Unfortunately, this didn't work (not even as part of `activate.bat`) - assuming I didn't make a mistake (non-negligible probability for that)...

**Comment by mingwandroid** [comment]
I will take a look.

**Comment by h-vetinari** [comment]
@mingwandroid Any news on this? :)

**Comment by mingwandroid** [comment]
Will try to look today! Feel free to keep pinging me though.

**Comment by h-vetinari** [comment]
> @mingwandroid: Will try to look today! Feel free to keep pinging me though.

Ping! :)

**Comment by h-vetinari** [comment]
@conda-forge-admin, please rerender

**Comment by h-vetinari** [comment]
Will be able to revert https://github.com/conda-forge/cryptography-feedstock/pull/37/commits/c93b950693a5d3c79b6b68e1702059c0f0541db4 once https://github.com/conda-forge/cryptography-vectors-feedstock/pull/30 lands

**Comment by h-vetinari** [comment]
Failure on linux/osx is related to the (patch-)version mismatch between `cryptography` and `cryptography-vectors`, and thus not a problem for the moment (see previous comment).
```
=================================== FAILURES ===================================
_____________________________ test_vector_version ______________________________

    def test_vector_version():
>       assert cryptography.__version__ == cryptography_vectors.__version__
E       AssertionError: assert '2.9.2' == '2.9.1'
E         - 2.9.1
E         ?     ^
E         + 2.9.2
E         ?     ^

tests/test_utils.py:3753: AssertionError
=============================== warnings summary ===============================
```

Wrong openssl version still being linked on windows (actually, the openssl version being used is easy to see from the log of the pytest runner):
```
============================= test session starts =============================
platform win32 -- Python 3.8.2, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
OpenSSL: OpenSSL 1.1.1  11 Sep 2018
rootdir: %SRC_DIR%
plugins: hypothesis-5.10.3
collected 107077 items / 11 skipped / 107066 selected

tests\test_cryptography_utils.py ....                                    [  0%]
[...]
```

**Comment by h-vetinari** [comment]
I tried to add a test for the linked version (the current situation with windows was caught by the cryptography test suite, but if the system openssl had been newer, it wouldn't have been) - I would have liked this explicit to see from the test section of the recipe, but I don't know how to let conda build find a script from the recipe folder, so now I just use the default name for the test script.

The test works as intended locally and in the windows-jobs on azure, but for some reason, on linux/osx, `openssl version` does not work as a subprocess call:
```
FileNotFoundError: [Errno 2] No such file or directory: 'openssl version': 'openssl version'
```

**Comment by h-vetinari** [comment]
Here's the output of the new `run_test.py` on linux:
```
===== testing package: cryptography-2.9.2-py37hb09aad4_0 =====
running run_test.py
Version used by cryptography:
OpenSSL 1.1.1g  21 Apr 2020
Version in conda environment:
OpenSSL 1.1.1g  21 Apr 2020
===== cryptography-2.9.2-py37hb09aad4_0 OK =====
import: 'cryptography'
[...]
```

and on windows (there was a race condition between print & error so I added a `sleep(1)`):

```
===== testing package: cryptography-2.9.2-py36hef61171_0 =====
running run_test.py
Version used by cryptography:
OpenSSL 1.1.1  11 Sep 2018
Version in conda environment:
OpenSSL 1.1.1g  21 Apr 2020
Traceback (most recent call last):
  File "D:\bld\cryptography_1587656578882\test_tmp\run_test.py", line 23, in <module>
    assert linked_version == env_version
AssertionError
```

**Comment by h-vetinari** [comment]
@mingwandroid
It seems the `activation.bat` is not being executed in the test environment. Even with explicitly trying to copy the activation script to the conda environment (as desribed [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#activate-scripts)), this isn't being picked up:

An exploratory `os.environ['CONDA_DLL_SEARCH_MODIFICATION_ENABLE']` in `run_test.py` raises
```
Traceback (most recent call last):
  File "D:\bld\cryptography_1587730154703\test_tmp\run_test.py", line 19, in <module>
    print(f"CONDA_DLL_SEARCH_MODIFICATION_ENABLE={os.environ['CONDA_DLL_SEARCH_MODIFICATION_ENABLE']}")
  File "D:\bld\cryptography_1587730154703\_test_env\lib\os.py", line 669, in __getitem__
    raise KeyError(key) from None
KeyError: 'CONDA_DLL_SEARCH_MODIFICATION_ENABLE'
```

**Comment by h-vetinari** [comment]
@isuruf 
conda-smithy still complains about the `AND`-connected license identifier (same for `OR` in a previous commit):

> @conda-forge-linter: For **recipe**:
>
> * License is not an SPDX identifier (or a custom LicenseRef) nor an SPDX license expression.

In another [context](https://github.com/conda-forge/conda-smithy/pull/1220#issuecomment-612639047), you suggested that this should work? I've double-checked that all individual license identifiers are [correct](https://spdx.org/licenses/).

**Comment by h-vetinari** [comment]
@conda-forge-admin, please rerender

**Comment by h-vetinari** [comment]
OK, so I finally managed to pass `CONDA_DLL_SEARCH_MODIFICATION_ENABLE` through to the test environment (as visible by the lack of KeyError in the log) by copying the activation files in `bld.bat`, and it works, at least on python 3.8.

For 3.6. & 3.7, something is still off. The conda env has `libcrypto.lib` & `libssl.lib` available, but it seems they get overruled by `dll`s: 
```
WARNING (cryptography,Lib\site-packages\cryptography\hazmat\bindings\_openssl.cp37-win_amd64.pyd): Needed DSO C:\Windows\System32\libcrypto-1_1-x64.dll not found in any `
WARNING (cryptography,Lib\site-packages\cryptography\hazmat\bindings\_openssl.cp37-win_amd64.pyd): Needed DSO C:\Windows\System32\libssl-1_1-x64.dll not found in any CDT/compiler package, nor the whitelist?!
```
OTOH, that exact same warning also gets displayed for 3.8, yet there it's working.

As mentioned further up, the `cryptography` install script will pass `libraries=["libssl", "libcrypto"]` to [`cffi.FFI().set_source`](https://github.com/pyca/cryptography/blob/2.9.2/src/_cffi_src/utils.py#L72-L78), from where it's again [passed on](https://cffi.readthedocs.io/en/latest/cdef.html#ffibuilder-set-source-preparing-out-of-line-modules) `setuptools`, resp. `distutils`.

My original goal was to find out if - on windows - `distutils` will always prefer `.dll` (anywhere on the path) over the corresponding `.lib` (which would explain the behaviour), but I got too lost in the weeds of setuptools/distutils to figure this out. Now I have to wonder if this behaviour is a bug in cpython that got solved in 3.8?

**Comment by h-vetinari** [comment]
Some more sleuthing: I searched for `CONDA_DLL_SEARCH_MODIFICATION_ENABLE` in all of github, and it seems it's used nowhere else in conda-forge, except in a [patch](https://github.com/conda-forge/python-feedstock/blob/5078cd91bdb6a10929b20444eac2a9eec0108f94/recipe/patches/0022-Add-CondaEcosystemModifyDllSearchPath.patch) on the feedstock of python itself. 

What's more, the patch only got added for [3.8](https://github.com/conda-forge/python-feedstock/pull/300) (where CI is working now), so I'm wondering if this means that the conda builds of python 3.6 & 3.7 need that patch? @mingwandroid

**Comment by isuruf** [comment]
I don't think a package should set the env variable `CONDA_DLL_SEARCH_MODIFICATION_ENABLE`. It's okay for testing, but not for end users. If it was, that would have been enabled by default.
Other software vendors shouldn't install libraries to `C:\Windows\System32`.

**Comment by h-vetinari** [comment]
> @isuruf: I don't think a package should set the env variable `CONDA_DLL_SEARCH_MODIFICATION_ENABLE.` It's okay for testing, but not for end users. If it was, that would have been enabled by default.
Other software vendors shouldn't install libraries to `C:\Windows\System32`.

That's fair enough, but not something I knew. Also, even just for testing, we'd (likely) need changes in the conda-packaged python. At least https://github.com/conda-forge/python-feedstock/pull/345 would be ready to go, AFAICT. I'd be curious to see if merging that PR would unblock the 3.7 build here.

**Comment by isuruf** [comment]
For testing, if it helps, I'll accept a PR to https://github.com/conda-forge/conda-forge-ci-setup-feedstock/blob/master/recipe/run_conda_forge_build_setup_win.bat to delete those annoying DLLs.

**Comment by h-vetinari** [comment]
> @isuruf: For testing, if it helps, I'll accept a PR to https://github.com/conda-forge/conda-forge-ci-setup-feedstock/blob/master/recipe/run_conda_forge_build_setup_win.bat to delete those annoying DLLs.

I created a PR, but that would IMO just push the issue with `C:\Windows\System32` & `CONDA_DLL_SEARCH_MODIFICATION_ENABLE` under the rug again.

I was also wondering: It's fair enough that vendors shouldn't install into `C:\Windows\System32`, but what if an end user *does* have those libs there already? Then they'd have a (potentially) vulnerable environment, while conda happily tells them they're on the newest openssl-version.

**Comment by h-vetinari** [comment]
@mingwandroid answered [here](https://github.com/conda-forge/python-feedstock/pull/345#issuecomment-619445752), copying his response and answering in this thread:
> In general, a random openssl in that dir is likely to be incompatible Abi wise with ours so will fail to load.
>
> At the end of the day windows dll loading is really poor and insecure but it's also the only thing we have (until something like midipix is ready for prime time use).
>
> We could add a symbol to our openssl and force linking to that somehow whenever the header is included or something?
>
> If CDSME did only system32 avoidance I would not object to recommending we set it globally or defaulted it to enabled though really. It does address the system32 problem well I think.

___________________________

My response

> @mingwandroid: In general, a random openssl in that dir is likely to be incompatible Abi wise with ours so will fail to load.

In this case, this is an almost 2-year-old openssl within that dir (`1.1.1` without a letter), but that means they should be abi-compatible, and it does in fact "work".

> We could add a symbol to our openssl and force linking to that somehow whenever the header is included or something?

How about running an activation script for `cryptography`that checks whether `libcrypto.dll` and `libssl.dll` are present in `C:\Windows\System32`, and reacting appropriately (either raising or possibly setting `CONDA_DLL_SEARCH_MODIFICATION_ENABLE`). That way at least, the vast majorities of installs would not be affected.

**Comment by isuruf** [comment]
I'm fine with adding a warning to `openssl-feedstock` using an activate script.

**Comment by mingwandroid** [comment]
If it has a different filename then this problem is not the same as what we are talking about.. I have a theory but I'm not look at it until tomorrow now..

**Comment by isuruf** [comment]
It's not. All `1.1.[0-9][a-z]` series have the same name, but each new version has new functions.

**Comment by h-vetinari** [comment]
> @mingwandroid: If it has a different filename then this problem is not the same as what we are talking about.. I have a theory but I'm not look at it until tomorrow now..

The filenames *are* different, but just the extensions. `C:\Windows\System32` has `libcrypto.dll` & `libssl.dll`, whereas the conda-Lib has `libcrypto.lib` & `libssl.lib`.

**Comment by isuruf** [comment]
`.lib`'s are import libraries for the corresponding dynamic libraries `.dll`s. (static libraries also have the same extension `.lib`). The linker looks for `.lib` which tells the linker about the functions exported in the `.dll`. Both the conda package and the system have the same file `libcrypto-1_1-x64.dll`

**Comment by h-vetinari** [comment]
> @isuruf: I'm fine with adding a warning to `openssl` conda package using an activate script.

I just started building such an activation script for openssl, and I found the following:
```
(builder) D:\Dev\conda-forge\openssl-feedstock\recipe>where libcrypto*
C:\ProgramData\Miniconda3\envs\builder\Library\bin\libcrypto-1_1-x64.dll
C:\ProgramData\Miniconda3\envs\builder\Library\bin\libcrypto-1_1-x64.pdb
C:\ProgramData\Miniconda3\Library\bin\libcrypto-1_1-x64.dll
C:\ProgramData\Miniconda3\Library\bin\libcrypto-1_1-x64.pdb
C:\Windows\System32\libcrypto.dll
C:\Program Files\Git\mingw64\bin\libcrypto-1_1-x64.dll
```
In other words, there exists an ancient (March 2019; around the age of the windows install) `libcrypto.dll` in `C:\Windows\System32`. Interestingly, `libssl.dll` is not present, however:
```
(builder) D:\conda-forge\openssl-feedstock\recipe>where libssl*
C:\ProgramData\Miniconda3\envs\builder\Library\bin\libssl-1_1-x64.dll
C:\ProgramData\Miniconda3\envs\builder\Library\bin\libssl-1_1-x64.dll.c~.conda_trash
C:\ProgramData\Miniconda3\envs\builder\Library\bin\libssl-1_1-x64.pdb
C:\ProgramData\Miniconda3\Library\bin\libssl-1_1-x64.dll
C:\ProgramData\Miniconda3\Library\bin\libssl-1_1-x64.dll.c~.conda_trash
C:\ProgramData\Miniconda3\Library\bin\libssl-1_1-x64.pdb
C:\Program Files\Git\mingw64\bin\libssl-1_1-x64.dll
```

This means that (at least as things currently stand), the behaviour depends only on the presence of `libssl.dll`, since even  *without* `CONDA_DLL_SEARCH_MODIFICATION_ENABLE`, I get the correct behaviour despite the presence of `libcrypto.dll` in `C:\Windows\System32`: 
```
(builder) D:\Dev\conda-forge\openssl-feedstock\recipe>echo %CONDA_DLL_SEARCH_MODIFICATION_ENABLE%
%CONDA_DLL_SEARCH_MODIFICATION_ENABLE%

(builder) D:\Dev\conda-forge\openssl-feedstock\recipe>python
Python 3.7.6 | packaged by conda-forge | (default, Mar 23 2020, 22:22:21) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> from cryptography.hazmat.backends.openssl import backend
>>> backend.openssl_version_text()
'OpenSSL 1.1.1g  21 Apr 2020'
```

Below the fold is the conda environment in question:

<details>

```
(builder) D:\Dev\conda-forge\openssl-feedstock\recipe>conda list
# packages in environment at C:\ProgramData\Miniconda3\envs\builder:
#
# Name                    Version                   Build  Channel
beautifulsoup4            4.9.0            py37hc8dfbb8_0    conda-forge
blinker                   1.4                        py_1    conda-forge
boolean.py                3.7                        py_0    conda-forge
brotlipy                  0.7.0           py37h4ab8f01_1000    conda-forge
bzip2                     1.0.8                hfa6e2cd_2    conda-forge
ca-certificates           2020.4.5.1           hecc5488_0    conda-forge
certifi                   2020.4.5.1       py37hc8dfbb8_0    conda-forge
cffi                      1.14.0           py37ha419a9e_0    conda-forge
chardet                   3.0.4           py37hc8dfbb8_1006    conda-forge
conda                     4.8.3            py37hc8dfbb8_1    conda-forge
conda-build               3.19.2           py37hc8dfbb8_0    conda-forge
conda-forge-pinning       2020.04.24.08.48.19               0    conda-forge
conda-package-handling    1.6.0            py37h702c6c1_2    conda-forge
conda-smithy              3.6.17                     py_0    conda-forge
cryptography              2.8              py37hb32ad35_1    conda-forge
decorator                 4.4.2                      py_0    conda-forge
deprecated                1.2.8              pyh9f0ad1d_0    conda-forge
filelock                  3.0.10                     py_0    conda-forge
git                       2.26.2                        0    conda-forge
gitdb                     4.0.4                      py_0    conda-forge
gitpython                 3.1.1                      py_0    conda-forge
glob2                     0.7                        py_0    conda-forge
idna                      2.9                        py_1    conda-forge
isodate                   0.6.0                      py_1    conda-forge
jinja2                    2.11.2             pyh9f0ad1d_0    conda-forge
libarchive                3.3.3             h0c0e0cf_1008    conda-forge
libiconv                  1.15              hfa6e2cd_1006    conda-forge
liblief                   0.9.0                ha925a31_3
libxml2                   2.9.10               h9ce36c8_0    conda-forge
license-expression        1.2                        py_0    conda-forge
lz4-c                     1.9.2                h33f27b4_0    conda-forge
lzo                       2.10              hfa6e2cd_1000    conda-forge
m2w64-gcc-libgfortran     5.3.0                         6
m2w64-gcc-libs            5.3.0                         7
m2w64-gcc-libs-core       5.3.0                         7
m2w64-gmp                 6.1.0                         2
m2w64-libwinpthread-git   5.0.0.4634.697f757               2
markupsafe                1.1.1            py37h8055547_1    conda-forge
menuinst                  1.4.16                   py37_0    conda-forge
msrest                    0.6.10                     py_0    conda-forge
msys2-conda-epoch         20160418                      1
networkx                  2.4                        py_1    conda-forge
oauthlib                  3.0.1                      py_0    conda-forge
openssl                   1.1.1g               he774522_0    conda-forge
pip                       20.0.2                     py_2    conda-forge
pkginfo                   1.5.0.1                    py_0    conda-forge
psutil                    5.7.0            py37h8055547_1    conda-forge
py-lief                   0.9.0            py37ha925a31_3
pycosat                   0.6.3           py37h8055547_1004    conda-forge
pycparser                 2.20                       py_0    conda-forge
pycrypto                  2.6.1           py37h8055547_1004    conda-forge
pygithub                  1.43.6           py37hc8dfbb8_1    conda-forge
pyjwt                     1.7.1                      py_0    conda-forge
pyopenssl                 19.1.0                     py_1    conda-forge
pysocks                   1.7.1            py37hc8dfbb8_1    conda-forge
python                    3.7.6           h60c2a47_5_cpython    conda-forge
python-libarchive-c       2.9                      py37_0    conda-forge
python_abi                3.7                     1_cp37m    conda-forge
pytz                      2019.3                     py_0    conda-forge
pywin32                   227              py37hfa6e2cd_0    conda-forge
pyyaml                    5.3.1            py37h8055547_0    conda-forge
requests                  2.23.0             pyh8c360ce_2    conda-forge
requests-oauthlib         1.2.0                      py_0    conda-forge
ripgrep                   12.0.1               h301d43c_1    conda-forge
ruamel.yaml               0.16.6           py37h8055547_1    conda-forge
ruamel.yaml.clib          0.2.0            py37h8055547_1    conda-forge
ruamel_yaml               0.15.80         py37h8055547_1001    conda-forge
scrypt                    0.8.13           py37h90766aa_3    conda-forge
setuptools                46.1.3           py37hc8dfbb8_0    conda-forge
six                       1.14.0                     py_1    conda-forge
smmap                     3.0.2              pyh9f0ad1d_0    conda-forge
soupsieve                 1.9.4            py37hc8dfbb8_1    conda-forge
sqlite                    3.30.1               hfa6e2cd_0    conda-forge
toolz                     0.10.0                     py_0    conda-forge
tqdm                      4.45.0             pyh9f0ad1d_0    conda-forge
urllib3                   1.25.9                     py_0    conda-forge
vc                        14.1                 h869be7e_1    conda-forge
vs2015_runtime            14.16.27012          h30e32a0_2    conda-forge
vsts-python-api           0.1.22                     py_0    conda-forge
wheel                     0.34.2                     py_1    conda-forge
win_inet_pton             1.1.0                    py37_0    conda-forge
wincertstore              0.2                   py37_1003    conda-forge
wrapt                     1.12.1           py37h8055547_1    conda-forge
xz                        5.2.5                h2fa13f4_0    conda-forge
yaml                      0.2.4                he774522_0    conda-forge
zlib                      1.2.11            h2fa13f4_1006    conda-forge
zstd                      1.4.4                h9f78265_3    conda-forge
```

</details>

**Comment by isuruf** [comment]
Did you run this on an azure instance or an appveyor instance?

**Comment by isuruf** [comment]
DLLs with different names don't interfere with conda environment. `libcrypto.dll` without the version doesn't interfere with conda.

**Comment by h-vetinari** [comment]
> Did you run this on an azure instance or an appveyor instance?

This is my local windows laptop. No further DLL (like `C:\Windows\System32\libcrypto-1_1-x64.dll`) present.

**Comment by isuruf** [comment]
If they are not there, building this feedstock in your local setup should work fine.

**Comment by h-vetinari** [comment]
> If they are not there, building this feedstock in your local setup should work fine.

Fair enough, but that's not the main issue here, right?

**Comment by isuruf** [comment]
I don't understand what you are trying to say here. What is the issue?

**Comment by h-vetinari** [comment]
> I don't understand what you are trying to say here. What is the issue?

Primarily, getting this build unblocked, and secondarily, how to protect users who might have system-libs interfering with their environment. Whether or not I can build locally is a distant third IMO (but I happily admit that I hadn't fully understood which dll's get picked up from where and how).

**Comment by isuruf** [comment]
> Primarily, getting this build unblocked

https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/84

> and secondarily, how to protect users who might have system-libs interfering with their environment. 

Add an activate script to `openssl` conda package.

**Comment by h-vetinari** [comment]
I'm already following up on that...
Just showed what I found while building the activation script, because I (wrongly) thought that this might be relevant.

**Comment by isuruf** [comment]
Ah, okay. Since there was confusion about what happened, let me explain. This issue is there if and only if one or both of the two files, `C:\Windows\System32\libcrypto-1_1-x64.dll`, `C:\Windows\System32\libssl-1_1-x64.dll` are present. The path has to be exact.

**Comment by h-vetinari** [comment]
I took out the activation commit for now (still have it backed up). This should still pass though (assuming conda-forge/conda-forge-ci-setup-feedstock#84 has been picked up already).

**Comment by h-vetinari** [comment]
Alright, the windows builds are picking up the right openssl!

Travis failure is just that old chestnut:
```
docker: write /var/lib/docker/tmp/GetImageBlob281102498: no space left on device.
```

**Comment by h-vetinari** [comment]
@conda-forge-admin, please rerender

**Comment by h-vetinari** [comment]
@isuruf
This should be ready (barring the same old travis errors that just need a restart).

**Comment by jjhelmus** [comment]
@h-vetinari Thanks for seeing this through.  LGTM.  
I'm going to merge this even with the failing Travis CI.  Those jobs can be restarted until a non-full worker is found.

**Comment by h-vetinari** [comment]
Yeah, this one was tricker than usual. Thanks for the help & guidance to everyone, and particularly to @isuruf & @mingwandroid!

**Pull request merged**